libmobilecoin/include/chacha20_rng.h

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+#ifndef MC_CHACHA20_RNG_H_
+#define MC_CHACHA20_RNG_H_
+
+#include "common.h"
+
+/* ==================== ChaCha20Rng ==================== */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct _ChaCha20Rng ChaCha20Rng;
+
+/// Returns a new ChaCha20Rng instance initialized with the
+/// seed value provided by the u64 long_val parameter
+///
+/// # Arguments
+///
+/// * `long_val` - an unsigned 64 bit value to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+ChaCha20Rng* MC_NULLABLE mc_chacha20_rng_create_with_long(
+    uint64_t value,
+    McError* MC_NULLABLE * MC_NULLABLE out_error
+);
+
+/// Returns a new ChaCha20Rng instance initialized with the
+/// seed value provided by the bytes data, which must be at
+/// least 32 bytes (only the first 32 bytes will be used)
+///
+/// # Arguments
+///
+/// * `bytes` - 32 bytes of data to use as the rng seed
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+ChaCha20Rng* MC_NULLABLE mc_chacha20_rng_create_with_bytes(
+    const McBuffer* MC_NONNULL bytes,
+    McError* MC_NULLABLE * MC_NULLABLE out_error
+)
+MC_ATTRIBUTE_NONNULL(1);
+
+/// Returns the current word_pos of the ChaCha20Rng instance
+///
+/// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 16 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+bool mc_chacha20_rng_get_word_pos(
+    ChaCha20Rng* MC_NONNULL chacha20_rng,
+    const McBuffer* MC_NONNULL out_word_pos,
+    McError* MC_NULLABLE * MC_NULLABLE out_error
+)
+MC_ATTRIBUTE_NONNULL(1,2);
+
+/// Sets the current word_pos of the ChaCha20Rng instance
+///
+/// /// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+/// * `out_word_pos` - pointer to buffer of 128 bytes where the current
+///   chacha20_rng wordpos will be returned
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+bool mc_chacha20_rng_set_word_pos(
+    ChaCha20Rng* MC_NONNULL chacha20_rng,
+    const McBuffer* MC_NONNULL bytes,
+    McError* MC_NULLABLE * MC_NULLABLE out_error
+)
+MC_ATTRIBUTE_NONNULL(1,2);
+
+/// Returns the next random u64 value from the ChaCha20Rng
+///
+/// /// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+///
+/// # Errors
+///
+/// * `LibMcError::Poison`
+uint64_t mc_chacha20_rng_next_long(
+    ChaCha20Rng* MC_NONNULL chacha20_rng,
+    McError* MC_NULLABLE * MC_NULLABLE out_error
+)
+MC_ATTRIBUTE_NONNULL(1);
+
+/// Frees the ChaCha20Rng
+///
+/// # Preconditions
+/// 
+/// * The ChaCha20Rng is no longer in use
+/// 
+/// # Arguments
+///
+/// * `chacha20_rng` - must be a valid ChaCha20Rng
+bool mc_chacha20_rng_free(
+    ChaCha20Rng* MC_NONNULL chacha20_rng
+)
+MC_ATTRIBUTE_NONNULL(1);
```

Comment:
The convention seems to be to take one nullable arg and return void; this matches the cbindgen-generated declaration

```suggestion
void mc_chacha20_rng_free(ChaCha20Rng* MC_NULLABLE chacha20_rng;
```

Reply:
I mean... it can be changed to be nullable, but does that really make sense?  I think the others should actually be changed to be non-null....

@the-real-adammork  thoughts?

Reply:
The cbindgen doesn't seem to specify null/not null... just has FfiOptOwnedPtr.

Reply:
After Slack convo, going to stick with non-null as there's no valid use of free with a nullable value :)   Thanks for all the help Remoun 👍

Reply:
I mentioned in my review, I think that having `free()` take `NULL` is the generally expected behavior in C, at least on modern platforms/compilers.

In terms of what cbindgen is specifying, I assume the `Opt` in `FfiOptOwnedPtr` refers to `Option<Box<T>>`. Like, `Box<T>` can never be `NULL`, but `Option` is defined as:

```rust
pub enum Option<B> {
    None,
    Some(B),
}
```

Recent Rust is smart enough to know that `Option<Box<T>>::None` is a way to represent a nullable pointer without giving anyone the ability to dereference it when it's `NULL`, so that's how it puts it together.